Cookie\DomainMatchesTest: improve and stabilize

jrfnl opened the pull request:
### Cookie\DomainMatchesTest::testUrlMatchManuallySet(): refactor to data provider

### Cookie\DomainMatchesTest::testDomainMatch*(): split type check tests off to separate test method

... with a `TypeProviderHelper` based data provider to safeguard all invalid types will be rejected.

### Cookie\DomainMatchesTest::testDomainMatch*(): use named data provider

This commit:
* Adds descriptive names to each of the cases being tested via the data provider.
* Adds keys to the parameters passed for each test case to make it easier to understand the test case.
* Improves the punctuation for the inline documentation in the data provider.

### Cookie\DomainMatchesTest::testDomainMatch*(): add extra test cases

.. to safeguard the existing behaviour.

### Cookie\DomainMatchesTest::testDomainMatch*(): minor documentation improvements

### Cookie::domain_matches(): minor code simplifications [1]

Remove a few redundant calculations and a redundant function call.

### Cookie::domain_matches(): minor code simplifications [2]

Store the value of the repeated strlen calculation for re-use instead of calling the function multiple times.

jrfnl committed:
Cookie\DomainMatchesTest::testDomainMatch*(): split type check tests off to separate test method

... with a `TypeProviderHelper` based data provider to safeguard all invalid types will be rejected.

jrfnl committed:
Cookie\DomainMatchesTest::testDomainMatch*(): use named data provider

This commit:
* Adds descriptive names to each of the cases being tested via the data provider.
* Adds keys to the parameters passed for each test case to make it easier to understand the test case.
* Improves the punctuation for the inline documentation in the data provider.

jrfnl committed:
Cookie\DomainMatchesTest::testDomainMatch*(): minor documentation improvements

jrfnl committed:
Cookie\DomainMatchesTest::testDomainMatch*(): add extra test cases

.. to safeguard the existing behaviour.

jrfnl committed:
Cookie::domain_matches(): minor code simplifications [1]

Remove a few redundant calculations and a redundant function call.

jrfnl committed:
Cookie\DomainMatchesTest::testUrlMatchManuallySet(): refactor to data provider

schlessera commented on src/Cookie.php:
```diff
@@ -207,12 +207,12 @@ public function domain_matches($domain) {
 
-		if (substr($domain, -1 * strlen($cookie_domain)) !== $cookie_domain) {
+		if (substr($domain, -(strlen($cookie_domain))) !== $cookie_domain) {
 			// The cookie domain should be a suffix of the passed domain.
 			return false;
 		}
 
-		$prefix = substr($domain, 0, strlen($domain) - strlen($cookie_domain));
+		$prefix = substr($domain, 0, -strlen($cookie_domain));
```
Why is one occurrence of `strlen()` being put into parentheses while the other isn't?

Also, it might be good to calculate the `$prefix` only once.

schlessera replied:
NVM the second remark, it's not the same `substr()`.

jrfnl replied:
I've added an extra commit which stores the calculated value and then automatically gets rid of the extra set of parentheses as well.

jrfnl committed:
Cookie::domain_matches(): minor code simplifications [2]

Store the value of the repeated strlen calculation for re-use instead of calling the function multiple times.
